# Problem 1  Pick-and-Place Robot MDP Design

In this problem, we design a Markov Decision Process (MDP) for a robot arm performing a repetitive pick‑and‑place task. The goal is to learn movements that are **fast**, **smooth**, and **successful**, using reinforcement learning.

---

## 1. State Space (S)

The state must fully describe the robot’s physical configuration so that the next action depends only on the current state (Markov property). A suitable state representation is:

- **Joint angles** of each motor:  
  \( \theta_1, \theta_2, \dots, \theta_n \)
- **Joint angular velocities**:  
  \( \dot{\theta}_1, \dot{\theta}_2, \dots, \dot{\theta}_n \)
- **End‑effector pose** (position + orientation):  
  \( (x, y, z, \phi, \psi, \gamma) \)
- **Gripper status**: open / closed
- **Object status**: picked / not picked

Thus, a state can be written as:



\[
s = [\theta_1, \dots, \theta_n,\ \dot{\theta}_1, \dots, \dot{\theta}_n,\ \text{ee\_pose},\ \text{gripper},\ \text{object\_status}]
\]



This captures everything needed for optimal control and satisfies the Markov property.

---

## 2. Action Space (A)

To learn smooth and fast movements, the agent must control the motors directly. Actions include:

- **Torque commands** to each joint:  
  \( a = [\tau_1, \tau_2, \dots, \tau_n] \)
- **Gripper commands**: open / close

This allows the agent to learn continuous, fine‑grained control of the arm.

---

## 3. Reward Function (R)

The reward must encourage **speed**, **smoothness**, and **successful task completion**.

A suitable reward design:

| Situation | Reward |
|----------|--------|
| Successful placement of object | **+20** |
| Successful grasp | **+5** |
| Each time step (to encourage speed) | **−1** |
| High‑jerk or high‑torque movement | **−5** |
| Dropping the object | **−10** |
| Invalid or unsafe action | **−5** |

This reward structure encourages the robot to:

- Complete the task quickly  
- Move smoothly  
- Avoid unnecessary or jerky motions  
- Avoid dropping the object  

---

## 4. Why This Is an MDP

This problem fits the MDP framework because:

- **States** fully describe the robot’s configuration.
- **Actions** directly influence the next state.
- **Transition dynamics** depend only on the current state and action.
- **Rewards** provide feedback for each action.
- The robot interacts with the environment over **time**, forming episodes.

---

## 5. Talking Points 

### **Talking Point A  Key RL Feature
A key reinforcement learning idea in this problem is the MDP design itself. Before the robot can learn anything, we must define what it “sees” (states), what it can “do” (actions), and how we score its behavior (rewards). This structure is what allows the agent to eventually learn a policy  basically a strategy  that tells it how to move the arm smoothly and efficiently.

### **Talking Point B  Implementation & Testing Challenge**
The biggest challenge here is getting the reward function right. If I reward the robot too lightly for smooth motion, it might move fast but in a jerky, unsafe way. If I penalize it too heavily, it becomes overly cautious and slow. Finding the right balance between speed and smoothness is tricky, and small changes in the reward weights can completely change the robot’s behavior.

### **Talking Point C  Why This Is Reinforcement Learning**
This setup is classic reinforcement learning because the robot improves through trial and error. It takes an action, sees what happens, gets a reward, and updates its behavior over time. The learning loop - state - action - reward - next state - follows exactly what Sutton & Barto describe. There’s no labeled dataset telling the robot the “correct” movement; it has to discover good behavior by interacting with the environment.

---


# Problem 2  2×2 Gridworld: Two Iterations of Value Iteration (Step‑by‑Step)

This problem uses the 2×2 gridworld defined in the assignment:

- States:  
  \( S = \{s_1, s_2, s_3, s_4\} \)
- Actions:  
  \( A = \{\text{up}, \text{down}, \text{left}, \text{right}\} \)
- Initial policy:  
  \( \pi(\text{up} \mid s) = 1 \) for all states
- Rewards:  
  \( R(s_1)=5,\ R(s_2)=10,\ R(s_3)=1,\ R(s_4)=2 \)
- Transitions:  
  Valid moves are deterministic; invalid moves keep the agent in the same state.
- Discount factor:  
  \( \gamma = 1 \) (not given, so we assume 1)

We perform **two full iterations** of value iteration.



#  Iteration 0  Initial Values

We start with:



\[
V_0(s_1)=0,\quad V_0(s_2)=0,\quad V_0(s_3)=0,\quad V_0(s_4)=0
\]



---

#  Iteration 1  Compute \( V_1 \)

 **State \( s_1 \)**  
Grid position: top‑left.  
Action = up (invalid) → stays in \( s_1 \)



\[
V_1(s_1) = R(s_1) + \gamma V_0(s_1)
\]




\[
V_1(s_1) = 5 + 0 = \boxed{5}
\]





 **State \( s_2 \)**  
Grid position: top‑right.  
Action = up (invalid) → stays in \( s_2 \)



\[
V_1(s_2) = 10 + 0 = \boxed{10}
\]





 **State \( s_3 \)**  
Grid position: bottom‑left.  
Action = up → moves to \( s_1 \)



\[
V_1(s_3) = 1 + V_0(s_1)
\]




\[
V_1(s_3) = 1 + 0 = \boxed{1}
\]





 **State \( s_4 \)**  
Grid position: bottom‑right.  
Action = up → moves to \( s_2 \)



\[
V_1(s_4) = 2 + V_0(s_2)
\]




\[
V_1(s_4) = 2 + 0 = \boxed{2}
\]



---

  **Updated Value Function After Iteration 1**



\[
V_1 = \{5,\ 10,\ 1,\ 2\}
\]



---

#  Iteration 2  Compute \( V_2 \)

Now we use \( V_1 \) values.



 **State \( s_1 \)**  
Invalid up → stays in \( s_1 \)



\[
V_2(s_1) = 5 + V_1(s_1)
\]




\[
V_2(s_1) = 5 + 5 = \boxed{10}
\]





 **State \( s_2 \)**  
Invalid up → stays in \( s_2 \)



\[
V_2(s_2) = 10 + V_1(s_2)
\]




\[
V_2(s_2) = 10 + 10 = \boxed{20}
\]





**State \( s_3 \)**  
Up → moves to \( s_1 \)



\[
V_2(s_3) = 1 + V_1(s_1)
\]




\[
V_2(s_3) = 1 + 5 = \boxed{6}
\]





 **State \( s_4 \)**  
Up → moves to \( s_2 \)



\[
V_2(s_4) = 2 + V_1(s_2)
\]




\[
V_2(s_4) = 2 + 10 = \boxed{12}
\]



---

#  **Final Value Function After Iteration 2**



\[
V_2 = \{10,\ 20,\ 6,\ 12\}
\]



---

#  Talking Points 

### **Talking Point A  Key RL Feature**
This problem demonstrates **value iteration**, which repeatedly applies the Bellman optimality update to move the value function closer to \( V^* \). It highlights how RL uses dynamic programming to evaluate and improve policies.

### **Talking Point B  Implementation & Testing Challenge**
The main challenge is keeping track of **valid vs. invalid actions**. Because invalid actions keep the agent in the same state, a mistake here completely changes the value updates. Ensuring the transition logic matches the grid layout is essential.

### **Talking Point C  Why This Is Reinforcement Learning**
Value iteration is a core RL planning algorithm. It uses the Bellman equation to compute long‑term returns and improve the policy. Even though this example is small, it follows Sutton & Barto’s RL framework: states, actions, rewards, transitions, and iterative policy improvement.







Problem 2  Code (2x2 Gridworld verification)

In [2]:


import numpy as np

gamma = 1.0  # assignment uses gamma = 1

# Rewards
R = {
    "s1": 5,
    "s2": 10,
    "s3": 1,
    "s4": 2
}

# Transition model for UP only
# Invalid UP = stay in same state
P_up = {
    "s1": "s1",  # top row, can't go up
    "s2": "s2",
    "s3": "s1",  # moves up to s1
    "s4": "s2"   # moves up to s2
}

states = ["s1", "s2", "s3", "s4"]

# Iteration 0
V0 = {s: 0 for s in states}

# Iteration 1
V1 = {}
for s in states:
    s_next = P_up[s]
    V1[s] = R[s] + gamma * V0[s_next]

# Iteration 2
V2 = {}
for s in states:
    s_next = P_up[s]
    V2[s] = R[s] + gamma * V1[s_next]

print("V0 =", V0)
print("V1 =", V1)
print("V2 =", V2)


V0 = {'s1': 0, 's2': 0, 's3': 0, 's4': 0}
V1 = {'s1': 5.0, 's2': 10.0, 's3': 1.0, 's4': 2.0}
V2 = {'s1': 10.0, 's2': 20.0, 's3': 6.0, 's4': 12.0}



### What these numbers mean

- **Iteration 0 (V0):**  
  We start with all values equal to zero. This is just the initial guess before the algorithm begins updating anything.

- **Iteration 1 (V1):**  
  Each state’s value becomes equal to its immediate reward, because the previous values were all zero.  
  So the agent is basically saying: “Right now, the best I know is just the reward I get for being in this state.”

- **Iteration 2 (V2):**  
  Now the values increase because the agent starts considering **future rewards**.  
  Since the policy always chooses **UP**, each state looks at the value of the state it transitions to when moving up.

  - `s1` and `s2` stay in place when moving up, so their values double.  
  - `s3` moves up to `s1`, so it gets its own reward plus the value of `s1`.  
  - `s4` moves up to `s2`, so it gets its own reward plus the value of `s2`.

### Why the values grow

Value iteration is basically asking:

 “If I keep following this policy forever, how much total reward will I collect?”

Because the rewards are positive and the agent keeps looping upward, the values increase as the algorithm accounts for more future steps.

### Final takeaway

The output confirms that:

- The code matches the assignment’s transition rules  
- The manual calculations are correct  
- The value function is moving toward the long‑term expected returns under the “always move up” policy  



# Problem 3  5×5 Gridworld: Value Iteration and In‑Place Updates

In this problem, we revisit the 5×5 gridworld from Lecture 3 and use Value Iteration to find the optimal value function and policy. Then we implement an in‑place version of Value Iteration and compare both approaches.

---

## 1. Environment in Plain Language

The grid is 5×5:

- Each cell is a **state** identified by its row and column, e.g., `s0,0`, `s2,2`, `s4,4`.
- The **goal state** is `s4,4`. When the agent reaches this state, the episode ends and it receives a reward of **+10**.
- There are **grey states**: `s2,2`, `s3,0`, and `s0,4`. These are “bad” states with a reward of **−5**.
- All other non‑terminal states give a reward of **−1**, which encourages the agent to reach the goal quickly.
- Actions are the usual moves: up, down, left, right. If the agent tries to move outside the grid, it simply stays in the same state.

So the agent is basically trying to **navigate the grid**, avoid the grey states, and reach the goal as efficiently as possible.

---

## 2. Standard Value Iteration (Synchronous Updates)

In standard Value Iteration, we:

1. Start with all state values initialized to zero (or some default).
2. Repeatedly apply the Bellman optimality update:

   

\[
   V_{k+1}(s) = \max_a \sum_{s'} P(s' \mid s, a)\,[R(s') + \gamma V_k(s')]
   \]



3. Stop when the values stop changing significantly (i.e., the maximum difference between \( V_{k+1} \) and \( V_k \) is below a small threshold).

Intuitively, each iteration makes the value function “look further into the future.” Early iterations only see short‑term rewards; later iterations incorporate longer‑term consequences of actions.

At the end, we extract the **optimal policy** by choosing, for each state, the action that gives the highest expected return according to the final value function.

---

## 3. In‑Place Value Iteration (Asynchronous Updates)

In the in‑place version, we use **one single array** for the values:

- As we sweep through the states, we **immediately overwrite** \( V(s) \) with its new value.
- When we move on to the next state, we may already be using updated values from earlier states in the same sweep.

This doesn’t change the final answer (we still converge to the same \( V^* \) and optimal policy), but it can change **how fast** we get there in terms of iterations.

You can think of it like this:

- Standard VI: “Update everything based on the old picture, then replace the picture.”
- In‑place VI: “Update the picture as you go, using the freshest information available.”

---

## 4. What the Results Look Like (High‑Level)

After running both versions:

- The **optimal value function** \( V^* \) is highest near the goal state `s4,4`, and decreases as you move farther away.
- States that are “on the way” to the goal along a good path have better values than states that lead into grey cells or dead ends.
- The **optimal policy** typically points the agent:
  - **Toward the goal**, while
  - **Avoiding grey states**, even if they are closer in terms of steps.

When you reshape the value function into a 5×5 grid, you can literally see a “hill” forming around the goal, with the arrows of the policy pointing uphill.

---

## 5. Comparison: Standard vs In‑Place

- **Convergence behavior:**  
  Both methods converge to the **same** optimal value function and policy.
- **Speed (iterations):**  
  In‑place updates often need **fewer sweeps** over the state space because they reuse updated values immediately.
- **Complexity:**  
  Per iteration, both have similar computational cost, but in‑place can be more efficient in practice.
- **Memory:**  
  In‑place uses a single value array, while standard VI typically uses two (old and new).

In practice, if the environment is large, in‑place updates can be a nice performance win without changing the theoretical result.

---

## 6. Talking Points 

### Talking Point A  Key RL Feature
The key RL feature here is **Value Iteration**, which combines **policy evaluation** and **policy improvement** into one update. Each Bellman update pushes the value function closer to the optimal one, and the greedy policy with respect to that value function becomes better over time.

### Talking Point B  Implementation & Testing Challenge
A big challenge is making sure the **reward function and transitions** are implemented exactly as described: correct goal state, correct grey states, and correct handling of invalid moves. A small bug (like mislabeling a grey state or mixing up row/column indices) can completely change the value landscape and policy.

### Talking Point C  Why This Is Reinforcement Learning
This setup is RL because we are solving a **sequential decision‑making problem** with states, actions, rewards, and a value function. Even though we’re using a planning method (Value Iteration) with a known model, it still follows the RL framework from Sutton & Barto: we define an MDP, compute returns via Bellman updates, and derive a policy that maximizes long‑term reward.

---


In [3]:
import numpy as np

# ----------------- Environment -----------------

class GridWorld5x5:
    def __init__(self, gamma=0.9):
        self.rows = 5
        self.cols = 5
        self.gamma = gamma

        self.goal = (4, 4)
        self.grey_states = {(2, 2), (3, 0), (0, 4)}

        self.rewards = self._build_rewards()

    def _build_rewards(self):
        R = np.full((5, 5), -1.0)
        for g in self.grey_states:
            R[g] = -5.0
        R[self.goal] = 10.0
        return R

    def is_terminal(self, s):
        return s == self.goal

    def step(self, s, a):
        r, c = s

        if a == 0:   # right
            new = (r, min(c + 1, 4))
        elif a == 1: # down
            new = (min(r + 1, 4), c)
        elif a == 2: # left
            new = (r, max(c - 1, 0))
        elif a == 3: # up
            new = (max(r - 1, 0), c)
        else:
            new = s

        if self.is_terminal(s):
            new = s

        reward = self.rewards[new]
        return new, reward

# ----------------- Standard Value Iteration -----------------

class ValueIterationAgent:
    def __init__(self, env, theta=1e-4):
        self.env = env
        self.theta = theta
        self.V = np.zeros((5, 5))
        self.policy = np.zeros((5, 5), dtype=int)

    def run(self):
        iteration = 0

        while True:
            delta = 0
            V_new = np.copy(self.V)

            for r in range(5):
                for c in range(5):
                    s = (r, c)

                    if self.env.is_terminal(s):
                        V_new[s] = self.env.rewards[s]
                        continue

                    q_vals = []
                    for a in range(4):
                        s2, reward = self.env.step(s, a)
                        q_vals.append(reward + self.env.gamma * self.V[s2])

                    V_new[s] = max(q_vals)
                    delta = max(delta, abs(V_new[s] - self.V[s]))

            self.V = V_new
            iteration += 1

            if delta < self.theta:
                break

        self.extract_policy()
        return iteration

    def extract_policy(self):
        for r in range(5):
            for c in range(5):
                s = (r, c)

                if self.env.is_terminal(s):
                    self.policy[s] = -1
                    continue

                q_vals = []
                for a in range(4):
                    s2, reward = self.env.step(s, a)
                    q_vals.append(reward + self.env.gamma * self.V[s2])

                self.policy[s] = int(np.argmax(q_vals))

# ----------------- In-Place Value Iteration -----------------

class InPlaceValueIterationAgent:
    def __init__(self, env, theta=1e-4):
        self.env = env
        self.theta = theta
        self.V = np.zeros((5, 5))
        self.policy = np.zeros((5, 5), dtype=int)

    def run(self):
        iteration = 0

        while True:
            delta = 0

            for r in range(5):
                for c in range(5):
                    s = (r, c)

                    if self.env.is_terminal(s):
                        self.V[s] = self.env.rewards[s]
                        continue

                    q_vals = []
                    for a in range(4):
                        s2, reward = self.env.step(s, a)
                        q_vals.append(reward + self.env.gamma * self.V[s2])

                    new_v = max(q_vals)
                    delta = max(delta, abs(new_v - self.V[s]))
                    self.V[s] = new_v

            iteration += 1

            if delta < self.theta:
                break

        self.extract_policy()
        return iteration

    def extract_policy(self):
        for r in range(5):
            for c in range(5):
                s = (r, c)

                if self.env.is_terminal(s):
                    self.policy[s] = -1
                    continue

                q_vals = []
                for a in range(4):
                    s2, reward = self.env.step(s, a)
                    q_vals.append(reward + self.env.gamma * self.V[s2])

                self.policy[s] = int(np.argmax(q_vals))

# ----------------- Run Both Agents -----------------

env = GridWorld5x5(gamma=0.9)

agent_std = ValueIterationAgent(env)
iters_std = agent_std.run()

agent_inp = InPlaceValueIterationAgent(env)
iters_inp = agent_inp.run()

print("Standard VI iterations:", iters_std)
print("In-Place VI iterations:", iters_inp)

print("\nStandard V*:\n", agent_std.V)
print("\nIn-Place V*:\n", agent_inp.V)

print("\nStandard Policy:\n", agent_std.policy)
print("\nIn-Place Policy:\n", agent_inp.policy)


Standard VI iterations: 10
In-Place VI iterations: 10

Standard V*:
 [[ 3.8706101  5.411789   7.12421    9.0269    11.141    ]
 [ 5.411789   7.12421    9.0269    11.141     13.49     ]
 [ 7.12421    9.0269    11.141     13.49      16.1      ]
 [ 9.0269    11.141     13.49      16.1       19.       ]
 [11.141     13.49      16.1       19.        10.       ]]

In-Place V*:
 [[ 3.8706101  5.411789   7.12421    9.0269    11.141    ]
 [ 5.411789   7.12421    9.0269    11.141     13.49     ]
 [ 7.12421    9.0269    11.141     13.49      16.1      ]
 [ 9.0269    11.141     13.49      16.1       19.       ]
 [11.141     13.49      16.1       19.        10.       ]]

Standard Policy:
 [[ 0  0  0  1  1]
 [ 0  0  0  0  1]
 [ 0  1  0  0  1]
 [ 0  0  0  0  1]
 [ 0  0  0  0 -1]]

In-Place Policy:
 [[ 0  0  0  1  1]
 [ 0  0  0  0  1]
 [ 0  1  0  0  1]
 [ 0  0  0  0  1]
 [ 0  0  0  0 -1]]


Both Standard Value Iteration and In‑Place Value Iteration converged to the same optimal value function and same optimal policy, which is exactly what should happen. The value table forms a smooth increase toward the goal at (4,4), and the policy arrows mostly point right and down, guiding the agent efficiently to the goal while avoiding grey states. Both methods took 10 iterations in this grid, which is normal  sometimes in‑place is faster, but in this case they behaved the same.

# Problem 4  Off‑Policy Monte Carlo with Importance Sampling

In this problem, we estimate the value of a target policy using data generated from a different behavior policy. This is called **off‑policy Monte Carlo**, and we correct for the mismatch between policies using **importance sampling**.

The idea is simple:
- The behavior policy generates trajectories.
- The target policy is the one we actually want to evaluate.
- Importance sampling reweights each return so that the estimate is unbiased.

We compute:
1. **Ordinary Importance Sampling (OIS)**
2. **Weighted Importance Sampling (WIS)**

WIS is usually more stable because it normalizes the weights.


### Talking Point A  Key RL Feature
This problem highlights **off‑policy learning**, where we evaluate one policy using data from another. Importance sampling corrects for the mismatch, making the estimate unbiased.

### Talking Point B  Implementation Challenge
The hardest part is computing the **importance weights** correctly. If any target action has zero probability under the behavior policy, the weight becomes zero and the episode contributes nothing.

### Talking Point C  Why This Is Reinforcement Learning
The agent learns from **full episodes**, using returns and policy ratios to estimate value. This follows the Monte Carlo prediction framework from Sutton & Barto.


In [4]:
import numpy as np
import random

# ----------------- Simple Episodic Environment -----------------

class SimpleEpisodeEnv:
    """
    A tiny environment for demonstrating off-policy Monte Carlo.
    Episode ends when terminal state is reached.
    """

    def __init__(self):
        self.states = [0, 1, 2, 3]   # example states
        self.terminal = 3
        self.actions = [0, 1]       # left or right

    def step(self, state, action):
        """
        Deterministic transitions for simplicity.
        """
        if state == 0:
            return 1, -1
        if state == 1:
            return 2, -1
        if state == 2:
            return 3, +10   # terminal reward
        return 3, 0

    def is_terminal(self, s):
        return s == self.terminal

# ----------------- Off-Policy Monte Carlo -----------------

class OffPolicyMC:
    def __init__(self, env, gamma=1.0):
        self.env = env
        self.gamma = gamma

    def generate_episode(self, behavior_policy):
        """
        behavior_policy: dict mapping state -> action probabilities
        """
        episode = []
        state = 0

        while not self.env.is_terminal(state):
            probs = behavior_policy[state]
            action = np.random.choice([0,1], p=probs)
            next_state, reward = self.env.step(state, action)
            episode.append((state, action, reward))
            state = next_state

        return episode

    def importance_sampling(self, target_policy, behavior_policy, episodes=50):
        """
        Computes both ordinary and weighted IS estimates.
        """
        ordinary_returns = []
        weighted_returns = []
        cumulative_weight = 0
        weighted_sum = 0

        for _ in range(episodes):
            ep = self.generate_episode(behavior_policy)

            G = 0
            W = 1

            for (s, a, r) in reversed(ep):
                G = self.gamma * G + r
                W *= target_policy[s][a] / behavior_policy[s][a]

            ordinary_returns.append(W * G)

            cumulative_weight += W
            weighted_sum += W * G

            if cumulative_weight == 0:
                weighted_returns.append(0)
            else:
                weighted_returns.append(weighted_sum / cumulative_weight)

        return ordinary_returns, weighted_returns


In [5]:
env = SimpleEpisodeEnv()
mc = OffPolicyMC(env)

# Behavior policy (random)
behavior = {
    0: [0.5, 0.5],
    1: [0.5, 0.5],
    2: [0.5, 0.5]
}

# Target policy (always choose action 1)
target = {
    0: [0.0, 1.0],
    1: [0.0, 1.0],
    2: [0.0, 1.0]
}

ordinary, weighted = mc.importance_sampling(target, behavior, episodes=50)

print("Ordinary IS estimate:", np.mean(ordinary))
print("Weighted IS estimate:", weighted[-1])


Ordinary IS estimate: 5.12
Weighted IS estimate: 8.0


The ordinary importance sampling estimate (≈ 5.12) is quite noisy because it directly averages 
𝑊
⋅
𝐺
, and the importance weights 
𝑊
 can vary a lot between episodes. The weighted importance sampling estimate (8.0) is more stable because it normalizes by the total weight, so extreme weights don’t dominate as much. In practice, we usually prefer the weighted estimate, and here it gives a higher, more reliable estimate of the target policy’s value.